### PR TITLE
Fix consent asking issue

### PR DESCRIPTION
**Pull request by Wathsara**
There is an issue where consent is not asking for the user-attributes which were not in the scopes at first and added to scopes later. This PR will Solve that issue.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -2222,6 +2164,91 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
 
+    /**
+     * Filter requested claims based on OIDC claims and return the claims which includes in OIDC.
+     *
+     * @param claimsForApproval
+     * @param oauth2Params
+     * @param spTenantDomain
+     * @return
+     * @throws RequestObjectException
+     * @throws ClaimMetadataException
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2175]: Add param descriptios

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 2]
```diff
@@ -2222,6 +2164,91 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+     */
+    private List<ClaimMetaData> getRequestedOidcClaimsList(ConsentClaimsData claimsForApproval,
+                                                           OAuth2Parameters oauth2Params, String spTenantDomain)
+            throws RequestObjectException, ClaimMetadataException {
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2179]: Add new line

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -2222,6 +2164,91 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+        List<ClaimMetaData> requestedOidcClaimsList = new ArrayList<>();
+        List<String> localClaimsOfOidcClaims = new ArrayList<>();
+        List<String> localClaimsOfEssentialClaims = new ArrayList<>();
+
+        // Get the claims uri list of all the requested scopes. Eg:- country, email
+        List<String> claimListOfScopes = openIDConnectClaimFilter.getClaimsFilteredByOIDCScopes(oauth2Params.
+                getScopes(), spTenantDomain);
+
+        // Get the requested claims came through request object
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2188]: Add a full stop(.)

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -565,12 +571,15 @@ during post consent handling to get the original ConsentClaimsData object (Assum
                 getSSOConsentService().processConsent(approvedClaimIds, serviceProvider,
                         loggedInUser, value, false);
             }
-        } catch (OAuthSystemException | SSOConsentServiceException e) {
+
+        } catch (OAuthSystemException | SSOConsentServiceException | RequestObjectException |
+                ClaimMetadataException e) {
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:576]: Shall we seperatly catch the two exceptions and add specific error messages.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 5]
```diff
@@ -2222,6 +2164,91 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+        }
+
+        if (CollectionUtils.isNotEmpty(essentialRequestedClaims)) {
+            // Get the external claims relevant to all essential requested claims.
+            Set<ExternalClaim> externalClaimSetOfEssentialClaims = ClaimMetadataHandler.getInstance()
+                    .getMappingsFromOtherDialectToCarbon
+                            (OIDC_DIALECT, new HashSet<String>(essentialRequestedClaims), spTenantDomain);
+
+                    /* Get the locally mapped claims for all the external claims of essential claims.
+                    Eg:- http://wso2.org/claims/country, http://wso2.org/claims/emailaddress
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2232]: Check formatting

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 6]
```diff
@@ -2222,6 +2164,91 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+
+        // Get the list of user info's essential claims.
+        for (RequestedClaim requestedClaim : requestedClaimsOfUserInfo) {
+            if (requestedClaim.isEssential()) {
+                essentialRequestedClaims.add(requestedClaim.getName());
+            }
+        }
+
+        if (CollectionUtils.isNotEmpty(claimListOfScopes)) {
+            // Get the external claims relevant to all oidc scope claims and essential claims
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2212]:
```suggestion
            // Get the external claims relevant to all oidc scope claims and essential claims.
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 1 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2183]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 2 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2187]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Added.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 5 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 6 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 7]
```diff
@@ -2222,6 +2172,92 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
 
+    /**
+     * Filter requested claims based on OIDC claims and return the claims which includes in OIDC.
+     *
+     * @param claimsForApproval Consent required claims.
+     * @param oauth2Params OAuth parameters.
+     * @param spTenantDomain Tenant domain.
+     * @return List of claims.
+     * @throws RequestObjectException
+     * @throws ClaimMetadataException
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2183]: Check formatting.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 8]
```diff
@@ -2222,6 +2172,92 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
         return additionalQueryParam;
     }
 
+    /**
+     * Filter requested claims based on OIDC claims and return the claims which includes in OIDC.
+     *
+     * @param claimsForApproval Consent required claims.
+     * @param oauth2Params OAuth parameters.
+     * @param spTenantDomain Tenant domain.
+     * @return List of claims.
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2181]: Shall we
```suggestion
     * @return Requested OIDC claim list.
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 7 above)
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2183]: Shall we add description to exceptions as well, like: "If an error occurred while..."

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 9]
```diff
@@ -2222,6 +2172,92 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
         return additionalQueryParam;
     }
 
+    /**
+     * Filter requested claims based on OIDC claims and return the claims which includes in OIDC.
+     *
+     * @param claimsForApproval Consent required claims.
+     * @param oauth2Params OAuth parameters.
+     * @param spTenantDomain Tenant domain.
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2180]: Check formatting

**Review by GANGANI — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 9 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 7 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 8 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 10]
```diff
@@ -2222,6 +2172,92 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+            if (requestedClaim.isEssential()) {
+                essentialRequestedClaims.add(requestedClaim.getName());
+            }
+        }
+
+        if (CollectionUtils.isNotEmpty(claimListOfScopes)) {
+            // Get the external claims relevant to all oidc scope claims and essential claims.
+            Set<ExternalClaim> externalClaimSetOfOidcClaims = ClaimMetadataHandler.getInstance()
+                    .getMappingsFromOtherDialectToCarbon
+                            (OIDC_DIALECT, new HashSet<String>(claimListOfScopes), spTenantDomain);
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2224]: Seems line breaks are not in the proper format. please check other places as well

Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed

**Review by AnuradhaSK — APPROVED** [review]

**Review by IsuraD — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 11]
```diff
@@ -550,11 +550,18 @@ during post consent handling to get the original ConsentClaimsData object (Assum
                 ConsentClaimsData object.
              */
             ConsentClaimsData value = getConsentRequiredClaims(loggedInUser, serviceProvider, oauth2Params);
+            if (value != null) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:553]: explain the reason to recalculate the requestedOidcClaimsList in oidc level as a comment.

**Review by GANGANI — APPROVED** [review]

**Pull request merged**